Bug: projectService's defaultProject resolves to the wrong path when relative

### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Issue Description

In https://github.com/microsoft/TypeScript/pull/59645, I'm trying to enable the project service for TS. I hit a number of issues when doing this; one of them is I wasn't able to load the config, getting errors like:

```
Parsing error: Could not read default project './scripts/tsconfig.json': error TS18003: No inputs were found in config file '/home/jabaile/work/TypeScript/scripts/scripts/tsconfig.json'. Specified 'include' paths were '["**/*.mjs","**/*.cjs"]' and 'exclude' paths were '[]'
```

Note `/home/jabaile/work/TypeScript/scripts/scripts/tsconfig.json`, weird.

This comes down to the fact the working directory passed in to the config file loading is `path.dirname(defaultProject)`:

https://github.com/typescript-eslint/typescript-eslint/blob/07e6866a70becf134c230acbe0ae5a5c4ae9850d/packages/typescript-estree/src/create-program/createProjectService.ts#L130-L134

So, if I pass in `./scripts/tsconfig.json`, it sets cwd to `./scripts`, then TS says "oh, a relative path, okay" and then tries to load `./scripts/scripts/tsconfig.json`.

The workaround is to give ts-eslint an absolute path, but I think the bug here is that it should actually be using `tsconfigRootDir` as the cwd.

### Reproduction Repository Link

https://github.com/microsoft/TypeScript/pull/59645

### Repro Steps

1) Clone https://github.com/microsoft/TypeScript/pull/59645
1) `npm ci`
1) Replace the defaultProject with `./scripts/tsconfig.json`
1) `npx eslint Herebyfile.mjs`


### Versions

<!-- Fill in the relevant versions and delete the others -->
| package                                | version |
| -------------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin`     | `8.1.0` |
| `@typescript-eslint/parser`            | `8.1.0` |
| `@typescript-eslint/scope-manager`     | `8.1.0` |
| `@typescript-eslint/typescript-estree` | `8.1.0` |
| `@typescript-eslint/type-utils`        | `8.1.0` |
| `@typescript-eslint/utils`             | `8.1.0` |
| `TypeScript`                           | `5.5.4` |
| `ESLint`                               | `9.9.0` |
| `node`                                 | `22.6.0` |

--- a/packages/typescript-estree/src/create-program/createProjectService.ts
+++ b/packages/typescript-estree/src/create-program/createProjectService.ts
@@ -1,6 +1,4 @@
 /* eslint-disable @typescript-eslint/no-empty-function -- for TypeScript APIs*/
-import path from 'node:path';
-
 import debug from 'debug';
 import type * as ts from 'typescript/lib/tsserverlibrary';
 
@@ -41,6 +39,7 @@ export interface ProjectServiceSettings {
 export function createProjectService(
   optionsRaw: boolean | ProjectServiceOptions | undefined,
   jsDocParsingMode: ts.JSDocParsingMode | undefined,
+  tsconfigRootDir: string | undefined,
 ): ProjectServiceSettings {
   const options = typeof optionsRaw === 'object' ? optionsRaw : {};
   validateDefaultProjectForFilesGlob(options);
@@ -130,7 +129,7 @@ export function createProjectService(
       configFile = getParsedConfigFile(
         tsserver,
         options.defaultProject,
-        path.dirname(options.defaultProject),
+        tsconfigRootDir,
       );
     } catch (error) {
       throw new Error(

--- a/packages/typescript-estree/src/parseSettings/createParseSettings.ts
+++ b/packages/typescript-estree/src/parseSettings/createParseSettings.ts
@@ -110,6 +110,7 @@ export function createParseSettings(
         ? (TSSERVER_PROJECT_SERVICE ??= createProjectService(
             tsestreeOptions.projectService,
             jsDocParsingMode,
+            tsconfigRootDir,
           ))
         : undefined,
     range: tsestreeOptions.range === true,

--- a/packages/typescript-estree/tests/lib/createProjectService.test.ts
+++ b/packages/typescript-estree/tests/lib/createProjectService.test.ts
@@ -46,13 +46,17 @@ describe('createProjectService', () => {
 
   it('sets allowDefaultProject when options.allowDefaultProject is defined', () => {
     const allowDefaultProject = ['./*.js'];
-    const settings = createProjectService({ allowDefaultProject }, undefined);
+    const settings = createProjectService(
+      { allowDefaultProject },
+      undefined,
+      undefined,
+    );
 
     expect(settings.allowDefaultProject).toBe(allowDefaultProject);
   });
 
   it('does not set allowDefaultProject when options.allowDefaultProject is not defined', () => {
-    const settings = createProjectService(undefined, undefined);
+    const settings = createProjectService(undefined, undefined, undefined);
 
     expect(settings.allowDefaultProject).toBeUndefined();
   });
@@ -68,6 +72,7 @@ describe('createProjectService', () => {
           allowDefaultProject: ['file.js'],
           defaultProject: './tsconfig.json',
         },
+        undefined,
         undefined,
       ),
     ).toThrow(
@@ -89,6 +94,7 @@ describe('createProjectService', () => {
           defaultProject: './tsconfig.json',
         },
         undefined,
+        undefined,
       ),
     ).toThrow(
       "Could not read default project './tsconfig.json': `getParsedConfigFile` is only supported in a Node-like environment.",
@@ -105,6 +111,7 @@ describe('createProjectService', () => {
         defaultProject: './tsconfig.json',
       },
       undefined,
+      undefined,
     );
 
     expect(service.setCompilerOptionsForInferredProjects).toHaveBeenCalledWith(
@@ -112,9 +119,34 @@ describe('createProjectService', () => {
     );
   });
 
+  it('uses tsconfigRootDir as getParsedConfigFile projectDirectory when provided', () => {
+    const compilerOptions = { strict: true };
+    const tsconfigRootDir = 'path/to/repo';
+    mockGetParsedConfigFile.mockReturnValue({ options: compilerOptions });
+
+    const { service } = createProjectService(
+      {
+        allowDefaultProject: ['file.js'],
+        defaultProject: './tsconfig.json',
+      },
+      undefined,
+      tsconfigRootDir,
+    );
+
+    expect(service.setCompilerOptionsForInferredProjects).toHaveBeenCalledWith(
+      compilerOptions,
+    );
+    expect(mockGetParsedConfigFile).toHaveBeenCalledWith(
+      // eslint-disable-next-line @typescript-eslint/no-require-imports
+      require('typescript/lib/tsserverlibrary'),
+      './tsconfig.json',
+      tsconfigRootDir,
+    );
+  });
+
   it('uses the default projects error debugger for error messages when enabled', () => {
     jest.spyOn(process.stderr, 'write').mockImplementation();
-    const { service } = createProjectService(undefined, undefined);
+    const { service } = createProjectService(undefined, undefined, undefined);
     debug.enable('typescript-eslint:typescript-estree:tsserver:err');
     const enabled = service.logger.loggingEnabled();
     service.logger.msg('foo', ts.server.Msg.Err);
@@ -130,7 +162,7 @@ describe('createProjectService', () => {
 
   it('does not use the default projects error debugger for error messages when disabled', () => {
     jest.spyOn(process.stderr, 'write').mockImplementation();
-    const { service } = createProjectService(undefined, undefined);
+    const { service } = createProjectService(undefined, undefined, undefined);
     const enabled = service.logger.loggingEnabled();
     service.logger.msg('foo', ts.server.Msg.Err);
 
@@ -140,7 +172,7 @@ describe('createProjectService', () => {
 
   it('uses the default projects info debugger for info messages when enabled', () => {
     jest.spyOn(process.stderr, 'write').mockImplementation();
-    const { service } = createProjectService(undefined, undefined);
+    const { service } = createProjectService(undefined, undefined, undefined);
     debug.enable('typescript-eslint:typescript-estree:tsserver:info');
     const enabled = service.logger.loggingEnabled();
     service.logger.info('foo');
@@ -156,7 +188,7 @@ describe('createProjectService', () => {
 
   it('does not use the default projects info debugger for info messages when disabled', () => {
     jest.spyOn(process.stderr, 'write').mockImplementation();
-    const { service } = createProjectService(undefined, undefined);
+    const { service } = createProjectService(undefined, undefined, undefined);
     const enabled = service.logger.loggingEnabled();
     service.logger.info('foo');
 
@@ -166,7 +198,7 @@ describe('createProjectService', () => {
 
   it('uses the default projects perf debugger for perf messages when enabled', () => {
     jest.spyOn(process.stderr, 'write').mockImplementation();
-    const { service } = createProjectService(undefined, undefined);
+    const { service } = createProjectService(undefined, undefined, undefined);
     debug.enable('typescript-eslint:typescript-estree:tsserver:perf');
     const enabled = service.logger.loggingEnabled();
     service.logger.perftrc('foo');
@@ -182,7 +214,7 @@ describe('createProjectService', () => {
 
   it('does not use the default projects perf debugger for perf messages when disabled', () => {
     jest.spyOn(process.stderr, 'write').mockImplementation();
-    const { service } = createProjectService(undefined, undefined);
+    const { service } = createProjectService(undefined, undefined, undefined);
     const enabled = service.logger.loggingEnabled();
     service.logger.perftrc('foo');
 
@@ -191,7 +223,7 @@ describe('createProjectService', () => {
   });
 
   it('enables all log levels for the default projects logger', () => {
-    const { service } = createProjectService(undefined, undefined);
+    const { service } = createProjectService(undefined, undefined, undefined);
 
     expect(service.logger.hasLevel(ts.server.LogLevel.terse)).toBe(true);
     expect(service.logger.hasLevel(ts.server.LogLevel.normal)).toBe(true);
@@ -200,7 +232,7 @@ describe('createProjectService', () => {
   });
 
   it('does not return a log filename with the default projects logger', () => {
-    const { service } = createProjectService(undefined, undefined);
+    const { service } = createProjectService(undefined, undefined, undefined);
 
     expect(service.logger.getLogFileName()).toBeUndefined();
   });
@@ -209,7 +241,7 @@ describe('createProjectService', () => {
     jest.spyOn(process.stderr, 'write').mockImplementation();
 
     debug.enable('typescript-eslint:typescript-estree:tsserver:event');
-    createProjectService(undefined, undefined);
+    createProjectService(undefined, undefined, undefined);
     debug.disable();
 
     expect(process.stderr.write).toHaveBeenCalledWith(
@@ -222,7 +254,7 @@ describe('createProjectService', () => {
   it('does not use the default projects event debugger for event handling when disabled', () => {
     jest.spyOn(process.stderr, 'write').mockImplementation();
 
-    createProjectService(undefined, undefined);
+    createProjectService(undefined, undefined, undefined);
 
     expect(process.stderr.write).toHaveBeenCalledTimes(0);
   });
@@ -232,6 +264,7 @@ describe('createProjectService', () => {
       {
         allowDefaultProject: ['file.js'],
       },
+      undefined,
       undefined,
     );
 

